File: 10/solution.ml
open Core
open Stdio

let read_input () = In_channel.read_lines "10/input.txt"

let parse_line line =
  let parts = String.split line ~on:' ' in
  let slights = List.hd_exn parts in
  let slights = String.sub slights ~pos:1 ~len:(String.length slights - 2) in

  (* lights to bitmask *)
  let lights =
    String.foldi slights ~init:0 ~f:(fun i acc c ->
        if Char.equal c '#' then acc lor (1 lsl i) else acc)
  in

  let sbuttons = List.slice parts 1 (List.length parts - 1) in
  let buttons =
    List.map sbuttons ~f:(fun s ->
        String.sub s ~pos:1 ~len:(String.length s - 2)
        |> String.split ~on:',' |> List.map ~f:Int.of_string)
  in

  (* buttons -> bitmasks *)
  let button_masks =
    List.map buttons ~f:(fun btn ->
        List.fold btn ~init:0 ~f:(fun acc i -> acc lor (1 lsl i)))
  in

  let sjolt = List.last_exn parts in
  let joltages =
    String.sub sjolt ~pos:1 ~len:(String.length sjolt - 2)
    |> String.split ~on:',' |> List.map ~f:Int.of_string
  in
  (lights, buttons, button_masks, joltages)

let min_presses target buttons =
  let memo = Hashtbl.create (module Int) in
  let queue = Queue.create () in
  Hashtbl.set memo ~key:0 ~data:0;
  Queue.enqueue queue 0;

  let rec bfs () =
    match Queue.dequeue queue with
    | None -> -1
    | Some state when state = target -> Hashtbl.find_exn memo state
    | Some state ->
        let presses = Hashtbl.find_exn memo state in
        List.iter buttons ~f:(fun btn ->
            let next = state lxor btn in
            if not (Hashtbl.mem memo next) then begin
              Hashtbl.set memo ~key:next ~data:(presses + 1);
              Queue.enqueue queue next
            end);
        bfs ()
  in
  bfs ()

let min_presses_ilp buttons joltages =
  let open Lp in
  let n_buttons = List.length buttons in

  (* x_i = press count for button i, integer >= 0 *)
  let vars =
    Array.init n_buttons ~f:(fun i ->
        var ~integer:true ~lb:0. (sprintf "x%d" i))
  in

  (* minimize total presses *)
  let objective =
    minimize (Array.fold vars ~init:(c 0.) ~f:(fun acc v -> acc ++ v))
  in

  (* each counter must equal its target  *)
  let constrs =
    List.mapi joltages ~f:(fun counter_idx target_val ->
        let lhs =
          Array.foldi vars ~init:(c 0.) ~f:(fun btn_idx acc v ->
              let btn = List.nth_exn buttons btn_idx in
              if List.mem btn counter_idx ~equal:Int.equal then acc ++ v
              else acc)
        in
        lhs =~ c (Float.of_int target_val))
  in

  let problem = make ~name:"joltage" objective constrs in
  match Lp_glpk.solve ~term_output:false problem with
  | Ok (obj, _) -> Float.iround_nearest_exn obj
  | Error _ -> -1

let part1 lines =
  List.fold lines ~init:0 ~f:(fun acc line ->
      let lights, _, button_masks, _ = parse_line line in
      acc + min_presses lights button_masks)

let part2 lines =
  List.fold lines ~init:0 ~f:(fun acc line ->
      let _, buttons, _, joltages = parse_line line in
      acc + min_presses_ilp buttons joltages)

let () =
  let lines = read_input () in
  printf "Part 1: %d\n" (part1 lines);
  printf "Part 2: %d\n" (part2 lines)

File: 10/solution.ts
import { type Constraint, solve, type Model } from "yalps";

const input = await Bun.file(`${import.meta.dir}/input.txt`).text();
const lines = input.trim().split("\n");

// create bitmasks for lights and buttons so we can xor state
function parseLine(line: string) {
	const parts = line.split(" ");
	const lights = parts
		.at(0)!
		.slice(1, -1)
		.split("")
		.reduce((acc, l, i) => acc | ((l === "#" ? 1 : 0) << i), 0);
	const buttons = parts
		.slice(1, -1)
		.map((b) => b.slice(1, -1).split(",").map(Number));
	const buttonMasks = buttons.map((b) =>
		b.reduce((acc, n) => acc | (1 << +n), 0),
	);
	const joltages = parts.at(-1)!.slice(1, -1).split(",").map(Number);

	return { lights, buttons, buttonMasks, joltages };
}

// bfs state -> min presses to reach from 0
function minPresses(target: number, buttons: number[]) {
	const memo = new Map<number, number>([[0, 0]]);
	const queue = [0];
	let head = 0;

	while (head < queue.length) {
		const state = queue[head++];
		if (state === target) return memo.get(state)!;

		for (const btn of buttons) {
			const next = state ^ btn;
			if (!memo.has(next)) {
				memo.set(next, memo.get(state)! + 1);
				queue.push(next);
			}
		}
	}

	return -1;
}

// integer linear programming
// joltages create a system of linear equations
// we need to find press counts such that:
// x1*btn1 + x2*btn2 ... xn*btn = joltage_target
function minPressesILP(buttons: number[][], joltages: number[]) {
	const variables: Record<string, Record<string, number>> = {};
	const constraints: Record<string, Constraint> = {};

	// each button is a variable
	for (let b = 0; b < buttons.length; b++) {
		const effect: Record<string, number> = { presses: 1 };
		for (const i of buttons[b]) {
			effect[`c${i}`] = 1;
		}

		variables[`x${b}`] = effect;
	}

	// each counter must hit exact target
	for (let i = 0; i < joltages.length; i++) {
		constraints[`c${i}`] = { equal: joltages[i] };
	}

	const model: Model = {
		direction: "minimize",
		objective: "presses",
		constraints,
		variables,
		integers: Object.keys(variables),
	};

	const result = solve(model);
	return result.status === "optimal" ? Math.round(result.result) : -1;
}

function part1() {
	return lines.reduce((sum, line) => {
		const { lights, buttonMasks } = parseLine(line);
		return sum + minPresses(lights, buttonMasks);
	}, 0);
}

function part2() {
	return lines.reduce((sum, line) => {
		const { joltages, buttons } = parseLine(line);
		return sum + minPressesILP(buttons, joltages);
	}, 0);
}

console.log("Part 1:", part1());
console.log("Part 2:", part2());
